File: quadratic_equation_solver.py
def read_the_value_from_the_user(variable):
    wrong_input=True
    while wrong_input==True:
        try:
            user_input=float(input(f"insert the value of {variable}"))
        except ValueError:
            pass
        else:
            wrong_input=False
    return user_input
variable_a=read_the_value_from_the_user("a")
variable_b=read_the_value_from_the_user("b")
variable_c=read_the_value_from_the_user("c")
def Calculate_the_discriminant():
    discriminant=(variable_b*variable_b)-(4*variable_a*variable_c)
    return discriminant
def calculate_solution_case_2(discriminant):
    print("this equation has 2 solutions")
    first_solution=(-variable_b+pow(discriminant,0.5))/(2*variable_a)
    second_solution=(-variable_b-pow(discriminant,0.5))/(2*variable_a)
    solutions=[]
    solutions.append(first_solution)
    solutions.append(second_solution)
    return solutions
def calculate_solution_case_3():
    print("this equation has only one solution")
    solution=f"x= {-variable_b/(2*variable_a)}"
    return solution
def solve_the_equation():
    if (variable_a==0):
        if (variable_b==0):
            if (variable_c==0):
                return "all numbers are solutions to this equation"
            else:
                return "this equation has no solution"
        else:
            return f"x= {-variable_c/variable_b}"
    else:
        discriminant=Calculate_the_discriminant()
        print(f"discriminant= {discriminant}")
        if (discriminant <0):
            return "this equation has no solution"
        if (discriminant >0):
            solution=calculate_solution_case_2(discriminant)
            return f"x= {solution[0]}\nx= {solution[1]}"
        if (discriminant==0):
            solution=calculate_solution_case_3()
            return solution
def print_result():
    result=solve_the_equation()
    print(result)
print_result()

File: quadratic_equation_solver.cpp
#include <iostream>
#include <cmath>
#include <limits>
using namespace std;
float read_user_input_value(string variable)
{
    float float_number;
    cout << "enter the value of "<<variable<<endl;
    cin>>float_number;
    while (cin.fail())
    {
        cin.clear();
        cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
        cout<< "Invalid Input , try something else "<<endl;
        cin>>float_number;
    }
    return float_number;
}
void solve_eq_case_2(float &discriminant,float &variable_a,float &variable_b,float &variable_c)
{
    float first_solution=(-variable_b+pow(discriminant,0.5))/(2*variable_a);
    float second_solution=(-variable_b-pow(discriminant,0.5))/(2*variable_a);
    cout<<"the first solution is "<<first_solution<<"\n";
    cout<<"the second solution is "<<second_solution<<endl;
}
void solve_eq_case_3(float &variable_a,float &variable_b,float &variable_c)
{
    cout<<"the solution is "<<-variable_b/(2*variable_a)<<endl;
}
void solve_equation(float &variable_a,float &variable_b,float &variable_c)
{
    if (variable_a==0)
    {
        if (variable_b==0)
        {
            if (variable_c==0)
            {
                cout<<"all numbers are solutions to this equation"<<endl;
            }
            else
            {
                cout<<"this equation has no solution"<<endl;
            }
        }
        else
        {
            cout<<-variable_c/variable_b<<endl;
        }
    }
    else
    {
        float discriminant=(variable_b*variable_b)-(4*variable_a*variable_c);
        cout<<"discriminant = "<<discriminant<<endl;
        if (discriminant<0)
        {
            cout<<"this equation has no solution"<<endl;
        }
        else if (discriminant>0)
        {
            solve_eq_case_2(discriminant,variable_a,variable_b,variable_c);
        }
        else if (discriminant==0)
        {
            solve_eq_case_3(variable_a,variable_b,variable_c);
        }
    }
}
int main()
{
    float variable_a=read_user_input_value("a");
    float variable_b=read_user_input_value("b");
    float variable_c=read_user_input_value("c");
    solve_equation(variable_a,variable_b,variable_c);
    system("pause");
    return 0;
}